pkg/agent/rest.go

```diff
@@ -0,0 +1,128 @@
+)
+
+// NodeZeroRestClient is a struct to interact with the Agent Rest API that is on node zero.
+type NodeZeroRestClient struct {
+	Client     *client.AssistedInstall
+	ctx        context.Context
+	config     client.Config
+	NodeZeroIP string
+}
+
+// NewNodeZeroRestClient Initialize a new rest client to interact with the Agent Rest API on node zero.
+func NewNodeZeroRestClient(ctx context.Context, assetDir string) (*NodeZeroRestClient, error) {
+	restClient := &NodeZeroRestClient{}
+
+	assetStore, err := assetstore.NewStore(assetDir)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create asset store")
+	}
+	nmState := &manifests.NMStateConfig{}
+	if err := assetStore.Fetch(nmState); err != nil {
+		return nil, errors.Wrapf(err, "failed to fetch %s", nmState.Name())
+	}
+
+	NodeZeroIP, err := manifests.GetNodeZeroIP(nmState.Config)
+	if err != nil {
+		return nil, err
+	}
+	config := client.Config{}
+	config.URL = &url.URL{
+		Scheme: "http",
+		Host:   net.JoinHostPort(NodeZeroIP, "8090"),
+		Path:   client.DefaultBasePath,
+	}
+	client := client.New(config)
+
+	restClient.Client = client
+	restClient.ctx = ctx
+	restClient.config = config
+	restClient.NodeZeroIP = NodeZeroIP
+
+	return restClient, nil
+}
+
+// IsRestAPILive Determine if the Agent Rest API on node zero has initialized
+func (rest *NodeZeroRestClient) IsRestAPILive() (bool, error) {
+	// GET /v2/infraenvs
+	listInfraEnvsParams := installer.NewListInfraEnvsParams()
+	_, err := rest.Client.Installer.ListInfraEnvs(rest.ctx, listInfraEnvsParams)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// GetRestAPIServiceBaseURL Return the url of the Agent Rest API on node zero
+func (rest *NodeZeroRestClient) GetRestAPIServiceBaseURL() *url.URL {
+	return rest.config.URL
+}
+
+// GetInfraEnvEvents Return the event list for the provided infraEnvID from the Agent Rest API
+func (rest *NodeZeroRestClient) GetInfraEnvEvents(infraEnvID *strfmt.UUID) (models.EventList, error) {
+	listEventsParams := &events.V2ListEventsParams{InfraEnvID: infraEnvID}
+	clusterEventsResult, err := rest.Client.Events.V2ListEvents(rest.ctx, listEventsParams)
+	if err != nil {
+		return nil, err
+	}
+	return clusterEventsResult.Payload, nil
+}
+
+// getClusterID Return the cluster ID assigned by the Agent Rest API
+func (rest *NodeZeroRestClient) getClusterID() (*strfmt.UUID, error) {
+	// GET /v2/clusters and return first result
+	listClusterParams := installer.NewV2ListClustersParams()
+	clusterResult, err := rest.Client.Installer.V2ListClusters(rest.ctx, listClusterParams)
+	clusterList := clusterResult.Payload
+	if err != nil {
+		return nil, err
+	}
+	if len(clusterList) == 1 {
+		clusterID := clusterList[0].ID
+		return clusterID, nil
+	} else if len(clusterList) == 0 {
+		logrus.Debug("cluster is not registered in rest API")
+		return nil, nil
+	} else {
+		logrus.Debugf("number of clusters found: %d", len(clusterList))
```

Comment:
IMHO this should be considered as a critical error. I couldn't figure out a case where this scenario could be possible, if not for a erroneous condition